### PR TITLE
Pass components into script runner

### PR DESCRIPTION
**Pull request by YuanTingHsieh**
Allow custom components for ScriptRunner

### Description

- BaseScriptRunner allows for customization
- ScriptRunner will remain the old behavior

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/job_config/base_script_runner.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        please make sure the ClientAPILauncherExecutor is created using the matching pipe_id.
+
+        Args:
+            script (str): Script to run. For in-process must be a python script path. For ex-process can be any script support by `command`.
+            script_args (str): Optional arguments for script (appended to script).
+            launch_external_process (bool): Whether to launch the script in external process. Defaults to False.
+            command (str): If launch_external_process=True, command to run script (preprended to script). Defaults to "python3".
+            framework (str): Framework type to connfigure converter and params exchange formats. Defaults to FrameworkType.PYTORCH.
+            params_transfer_type (str): How to transfer the parameters. FULL means the whole model parameters are sent.
+                DIFF means that only the difference is sent. Defaults to TransferType.FULL.
```
Comment (yanchengnv) [inline nvflare/job_config/base_script_runner.py:72]: Please add description for other newly added args.

File: nvflare/job_config/script_runner.py [hunk 2]
```diff
@@ -12,12 +12,9 @@
...
 # limitations under the License.
 
-from typing import Type
+from nvflare.client.config import TransferType
 
-from nvflare.app_common.executors.client_api_launcher_executor import ClientAPILauncherExecutor
-from nvflare.app_common.executors.in_process_client_api_executor import InProcessClientAPIExecutor
-from nvflare.client.config import ExchangeFormat, TransferType
-from nvflare.fuel.utils.import_utils import optional_import
+from .base_script_runner import BaseScriptRunner
```
Comment (yanchengnv) [inline nvflare/job_config/script_runner.py:17]: Also import FrameworkType from BaseScriptRunner, and remove it from this file.

**Review by yanchengnv — CHANGES_REQUESTED** [review]
See my comments. You may want to combine both files into the same file script_runner.py.

File: nvflare/job_config/base_script_runner.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            raise ValueError(f"Framework {self._framework} unsupported")
+
+        if metric_pipe is not None:
+            validate_object_for_job("metric_pipe", metric_pipe, Pipe)
+        if metric_relay is not None:
+            validate_object_for_job("metric_relay", metric_relay, MetricRelay)
+        if task_pipe is not None:
+            validate_object_for_job("task_pipe", task_pipe, Pipe)
+        if executor is not None:
+            validate_object_for_job("executor", executor, Executor)
```
Comment (SYangster) [inline nvflare/job_config/base_script_runner.py:109]: maybe we could add a more specific check for executor being a `ClientAPILauncherExecutor` or `InProcessClientAPIExecutor` depending on `launch_external_process`

File: nvflare/job_config/script_runner.py [hunk 4]
```diff
@@ -60,126 +57,11 @@ def __init__(
 
         self._params_exchange_format = None
```
Comment (SYangster) [inline nvflare/job_config/script_runner.py:58]: we no longer need to set these instance variables here

File: nvflare/job_config/base_script_runner.py (hunk 1 above)
Comment (YuanTingHsieh) [reply nvflare/job_config/base_script_runner.py]: thanks, added.

File: nvflare/job_config/script_runner.py (hunk 2 above)
Comment (YuanTingHsieh) [reply nvflare/job_config/script_runner.py]: merge the two files as suggested.

File: nvflare/job_config/base_script_runner.py (hunk 3 above)
Comment (YuanTingHsieh) [reply nvflare/job_config/base_script_runner.py]: thanks! added as suggested.

**Review by yanchengnv — APPROVED** [review]
LGTM

**Pull request merged**